### PR TITLE
Feature/extended interpolate

### PR DESCRIPTION
**Pull request by SergeDemoulinGebit**
new ElementHandling: "extended_interpolate"

- Take the element from the interpolated POM, but resolve the variables with the properties from the effective POM.
- Don't resolve the following variables "project.basedir", "project.baseUri", "project.build.directory", "project.build.outputDirectory", "project.build.sourceDirectory", "project.build.scriptSourceDirectory", "project.build.testSourceDirectory", "project.reporting.outputDirectory"
- Interpolate the parent version

Optimization (only the used models are created)

### CODE REVIEW

**Comment by SergeDemoulinGebit** [comment]
resolve #383 resolve #382

**Comment by slawekjaranowski** [comment]
Please don't change line end and other whitespace

**Comment by SergeDemoulinGebit** [comment]
Hi
I fix the line ends now. 
Sorry for the late answer, I did not see the comment. This feature is important for us and very useful.

I want to remove the dependency to guava. Please wait until I do that to make the review.

Best regards,
Serge

**Comment by kjq** [comment]
This would be huge for us because we have a lot of properties in a parent that need to be interpolated in lower POMs.

**Comment by cpfeiffer** [comment]
Is there anything we can do to move this forward?

**Comment by slawekjaranowski** [comment]
I added myself to review, I will try to do it in a few days - @SergeDemoulinGebit thanks for PR

File: src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java [hunk 1]
```diff
@@ -447,6 +451,67 @@ protected boolean shouldSkipGoal() {
         return skipFlatten;
     }
 
+    private void logParameters() {
+
+        getLog().info("Generating flattened POM of project " + this.project.getId() + " with following parameters "
+                + this.asString(this.getParameters()));
+    }
```
Comment (slawekjaranowski) [inline src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java:458]: This method and corresponding is not needed we have all this values in debug log, executing Maven with -X
like:
```
ehaus.mojo:flatten-maven-plugin:1.6.1-SNAPSHOT:flatten from plugin realm ClassRealm[plugin>org.codehaus.mojo:flatten-maven-plugin:1.6.1-SNAPSHOT, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@63c12fb0]
[DEBUG] Configuring mojo execution 'org.codehaus.mojo:flatten-maven-plugin:1.6.1-SNAPSHOT:flatten:default-cli' with basic configurator -->
[DEBUG]   (f) defaultOperation = flatten
[DEBUG]   (f) embedBuildProfileDependencies = false
[DEBUG]   (f) flattenMode = bom
[DEBUG]   (f) flattenedPomFilename = .flattened-pom.xml
[DEBUG]   (f) keepCommentsInPom = false

```

**Comment by SergeDemoulinGebit** [comment]
right. I learnt it afterall

File: src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java [hunk 2]
```diff
@@ -23,41 +23,19 @@
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
-import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.io.StringWriter;
+import java.io.*;
```
Comment (slawekjaranowski) [inline src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java:26]: please don't use asterisk in imports

File: src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java [hunk 3]
```diff
@@ -447,6 +425,61 @@ protected boolean shouldSkipGoal() {
         return skipFlatten;
     }
 
+    private String asString(Map<String, Object> map) {
```
Comment (slawekjaranowski) [inline src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java:428]: where is used? and other new methods

**Review by slawekjaranowski — CHANGES_REQUESTED** [review]

**Comment by cpfeiffer** [comment]
Kind request -- can we move forward with this?

**Review by slawekjaranowski — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
@SergeDemoulinGebit thanks

**Comment by cpfeiffer** [comment]
Awesome, thank you so much!

**Comment by SergeDemoulinGebit** [comment]
Thank you Slawomir 😄
________________________________

**Pull request merged**